toonst/RIOT
# Tests/candev native4

### Contribution description
Adds a candev test app that can be used to control can-devices through the candev-abstraction, but without using the can-DLL.

Currently, only native is supported (because we thought it made sense to make a PR with only native first, so the main concepts can be tested, reviewed and commented on), but code that adds the MCP2515 driver (see #xxxx) to this test-app has already been written and tested, and will be made into a PR soon, along with the driver-code of this old PR. 


### Testing procedure
* Make sure libsocketcan and canutils are installed and running (see: https://github.com/RIOT-OS/RIOT/tree/master/tests/conn_can) TODO: maybe add this to the README of this test-app as well.
* Start candump on vcan0
* In the test-app shell: the `send` command will send a CAN message to vcan0. It should become visible in candump. Optionally you can pass some bytes (in decimal notation) as arguments (e.g. `send 12 34 56`). If no arguments are supplied, three default bytes will be sent. 
* To test the receive part, we can send data from Linux with cansend (e.g. `cansend vcan0 001#AABB`) or with cangen (e.g. `cangen vcan0 -v -n 1`).
* In the test-app, the received bytes will be stored in a FIFO buffer and can be retried by calling `receive`. (an optional argument n can be passed to receive multiple can-messages). If there are no CAN-messages in the buffer when `receive` is called, it will block until a message is received.
